### PR TITLE
Measure mode

### LINKED ISSUE #52: [FR] Add measurement tool
First thank you for your project, it's very clean.
Is it possible to add a tool to make measurement on the model :
1. Measuring the length of an edge
2. The distance between 2 parallel faces
3. The angle between 2 faces/edge
4. (Optionally the surface for a polygon)

### PR DESCRIPTION
**Pull request by X-Ryl669**
This adds a measurement tool to the application.
Usage is very intuitive:
1. Click on the measure icon in the toolbar
2. Select a face on the model
3. Select another face on the model
4. *Et voilà*, the measure appears on the left bar!

image

Closes #52.

### CODE REVIEW

**Comment by kovacsv** [comment]
Hi @X-Ryl669!

Thank you again for your contribution. Your solution seems mostly working, but unfortunately I feel that it doesn't fit in the current logic.

From the user experience perspective:
- The selection highlight circle's size doesn't depend on the size of the model. For small models it seems huge.
- You have two calculated values: "Distance", and "Length". For me these two words mean almost the same. I understand that length is the exact length of the clicked points, but from the current red circle feedback it's not clear which point is clicked.
- It's not clear for me why the "Distance" value is useful, and what does it exactly calculates.
- The measures tab is always visible, and gains space from other information even when measure mode is not active. I think that the measured values should be displayed somewhere else, and not in the left bar.

From the functionality perspective:
- When I turn measure mode off, the feedback elements disappear, that's ok. But it seems that it's still in measure mode, and the old functionality (selecting objects by clicking) doesn't work anymore.
- When you drag and drop another model, the measure mode keeps activated, but doesn't work.

image

So to summarize, I don't think I'd like to go in this direction with the measure tool.

**Comment by X-Ryl669** [comment]
> * The selection highlight circle's size doesn't depend on the size of the model. For small models it seems huge.

Yes, I forgot to account for this. I'll use the bounding box of the objec (world?) as a scale. Also, I think it's possible to draw a (transparent) cross as a texture on the circle so it more obvious what was the clicked point. 

 
> * You have two calculated values: "Distance", and "Length". For me these two words mean almost the same. I understand that length is the exact length of the clicked points, but from the current red circle feedback it's not clear which point is clicked.
 

Distance is the most useful value, it's the distance between 2 parallel planes (usually what you measure with a tape). It'll be zero when you measure between something else than a parallel planes (since the planes intersect, the distance is zero). Length is, as you guessed the distance between clicked points. I've initially called it "Between points" but it does not fit and the text takes 2 lines which is ugly. The red circle is an example, a target or a cross could be better.

> * It's not clear for me why the "Distance" value is useful, and what does it exactly calculates.


If you are to measure an object, you'll most likely need the distance since point selection is imprecise. The distance is, mathematically, the minimum distance between 2 points on the selected planes. 

> * The measures tab is always visible, and gains space from other information even when measure mode is not active. I think that the measured values should be displayed somewhere else, and not in the left bar.

I've used another [div](https://blog.cyril.by/en/documentation/3d%20models/filament-detector) in the embed code, but it seems more logical to use the "info" panel for information. It's probably possible to hide/remove the measure panel when it's not used.



> * When I turn measure mode off, the feedback elements disappear, that's ok. But it seems that it's still in measure mode, and the old functionality (selecting objects by clicking) doesn't work anymore.

No, we aren't in measure more anymore (the "destructor" of the FaceSelector is responsible for the button status and measure panel and then it's null-ified). But I had to modify the behavior for allowing measuring when on mobile browser (touch based) to mimic click, I've likely broken something here. To be honest, O3DV does not work well on mobile browser.
If you have an example model file where it does not work it'll be useful to post it here, so I can fix the behavior.

> * When you drag and drop another model, the measure mode keeps activated, but doesn't work.

TBH, I haven't tested with multiple models. Right now the measure mode is attached to a single mesh, so yes, it can't swap to another model in the middle of a measure. I guess it's a "Measure model" tool, not a "Measure world" tool. I don't really know what should be the correct behavior here. I think it's possible to make it work on multiple model, but I need to rework a bit. 
Again, please post a multi model file so I can test it's behaving as expected.

**Comment by X-Ryl669** [comment]
With the changes, I think I've resolved the issue you mentioned except for the measure panel where I need to know if hiding it would be ok for you or if it needs to go elsewhere.
image

**Comment by X-Ryl669** [comment]
Now it looks like this (example in embedded code):
image

**Comment by zhaoxiaomiao** [comment]
> 这为应用程序添加了一个测量工具。
> 用法非常直观：
> 
> 1. 单击工具栏中的度量图标
> 2. 在模型上选择一个面
> 3. 在模型上选择另一个面
> 4. _Et voilà_，该度量出现在左侧栏上！
> 
> 图像
> 
> 关闭 #52。

I think the function needs to be enhanced, such as increasing the distance between two points, measuring angle, measuring the area of the selected face, and angle, etc. I don't know if it can be added?

**Comment by X-Ryl669** [comment]
I've updated the code on my repository, so it's a bit better looking now (see above).
I'm currently thinking of adding the area the selected face and adding the volume of the selected solid (provided it's closed).
Measuring angles between 2 planes is done already (pick one plane, then pick another one, you'll get the angle). 

If you think about picking points or edges (instead of faces), then no, it's not available and would require a different logic because the models are not solid (in CAD terms), but meshes. So you'll get a lot of edges and vertices everywhere (even in the middle of a plane). 

For example, think of a cube. In a STL file, each face is made of 2 triangles, so you'll get 6 edges instead of 4. If you were to pick the edges, you'll have to filter the diagonal *virtual* edge, but this means sorting all vertices and matching their edges (a diagonal on a cube's face is shared by 2 edges and 2 vertices) and somehow figuring out that it's the same plane as the other set of 3 vertices, and then remove this edge from the picking set. 

Now think about a staircase on a floor plan. The floor plan will have (at least) 2 triangles, and the top stair will also have 2 triangles, yet the edge between the top stair and the floor plan is the same as the side of the floor plan, so if you remove it from the picking set, you can't select the floor plane edge. If you keep it, then you have a false edge between the top stair and the floor plan.

So as you see, there is no magical formula for measuring if you don't have the solid information (face, edge, volume) like a CAD file is making and only a mesh. You can do some processing (like I did) but not as much as you would do if you had a true object and a ruler.

**Comment by zhaoxiaomiao** [comment]
> 我已经更新了我的存储库中的代码，所以现在看起来更好（见上文）。
> 我目前正在考虑添加所选面的区域并添加所选实体的体积（前提是它已关闭）。
> 测量两个平面之间的角度已经完成（选择一个平面，然后选择另一个平面，您将获得角度）。
> 
> 如果您考虑选择点或边（而不是面），那么不，它不可用并且需要不同的逻辑，因为模型不是实体（在 CAD 术语中），而是网格。所以你会在任何地方（甚至在平面的中间）得到很多边和顶点。
> 
> 例如，考虑一个立方体。在 STL 文件中，每个面由 2 个三角形组成，因此您将获得 6 条边，而不是 4 条。如果要选取边，则必须过滤对角_虚拟_边，但这意味着对所有顶点进行排序并匹配它们的边（立方体面上的一条对角线由 2 个边和 2 个顶点共享）并以某种方式确定它与另一组 3 个顶点在同一平面上，然后从拾取集中删除这条边。
> 
> 现在想想平面图上的楼梯。平面图将有（至少）2 个三角形，顶部楼梯也将有 2 个三角形，但顶部楼梯和平面图之间的边缘与平面图的边相同，因此如果您将其从拾取集，您不能选择地板平面边缘。如果你保留它，那么顶部楼梯和平面图之间就会有一个错误的边缘。
> 
> 因此，如您所见，如果您没有像 CAD 文件制作的实体信息（面、边缘、体积）并且只有网格，则没有神奇的公式可以用于测量。你可以做一些处理（就像我做的那样），但不会像你有一个真实的物体和一把尺子那样做。

I agree with you. What you said is quite right. In regard to the way to measure this piece, I think we can refer to opencascade's Web 3D viewer. You can refer to it as follows:

https://www.opencascade.com/web-3d-viewer.php

**Comment by X-Ryl669** [comment]
OpenCascade's viewer is really impressive at first. But if you look closely you'll see that:
1. They only import solid geometry (OCC is the kernel in FreeCAD for example) so they have the solid information (we don't)
2. The measure tool is not precise (you pick a point and another point, but you don't exactly know where you pick the point, and nothing more). The PR here (or better, checkout my `merged` branch), does the same and more, it compute also the distance & angle between different planes (if you pick points on different planes)
3. They have a very useful cut tool (IMHO) and I think it's possible to create the same with ThreeJS, see [here](https://threejs.org/examples/?q=clipping#webgl_clipping_stencil). It up to us to add it.
4. They have a pseudo static navigation cube (I'm working on this right now, hopefully, it'll be better than theirs)
5. They don't compute the easy measures of the objects (volume & surface), while it's already in O3DV.
6. They have a very nice UI that's intuitive, O3DV on the opposite is not very intuitive unless using the embedded player and making your own interface, see [here](https://blog.cyril.by/fr/documentation/3d%20models/vga-connector).
7. They are a company with a lot of resources, we are individual developers 

All in all, I'd say that O3DV is a very nice and solid software and only few things are missing. Hopefully, it'll be added in the future.

**Comment by kovacsv** [comment]
A measure tool is implemented now in release 0.8.8.
